sdk/storage/Azure.Storage.Blobs/src/PartitionedUploader.cs

```diff
@@ -0,0 +1,249 @@
+
+                if (runningTasks.Count >= _threadCount)
+                {
+                    await Task.WhenAny(runningTasks).ConfigureAwait(false);
+
+                    runningTasks.RemoveAll(t => t.IsCompleted);
+                }
+            }
+
+            await Task.WhenAll(runningTasks).ConfigureAwait(false);
+
+            return await _client.CommitBlockListAsync(blocks, blobHttpHeaders, metadata, blobAccessConditions, accessTier, cancellationToken)
+                .ConfigureAwait(false);
+        }
+
+        private async Task StageBlobAsync(StreamPartition block, string blockId, BlobRequestConditions conditions, IProgress<long> progressHandler, CancellationToken cancellationToken)
+        {
+            try
+            {
+                await _client.StageBlockAsync(
+                    blockId,
+                    new MemoryStream(block.Bytes, 0, block.Length, writable: false),
+                    conditions: conditions,
+                    progressHandler: progressHandler,
+                    cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                block.Dispose();
+            }
+        }
+
+        private static bool TryGetLength(Stream content, out long length)
+        {
+            length = 0;
+            try
+            {
+                if (content.CanSeek)
+                {
+                    length = content.Length;
+                    return true;
+                }
+            }
+            catch (NotSupportedException)
+            {
+            }
+            return false;
+        }
+
+        private static string GenerateBlockId(long offset)
+        {
+            byte[] id = new byte[48];
+
+            BitConverter.GetBytes(offset).CopyTo(id, 0);
+
+            return Convert.ToBase64String(id);
+        }
+
+        private async IAsyncEnumerable<StreamPartition> GetBlocksAsync(Stream stream, bool async, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            int read;
+            long absolutePosition = 0;
+            do
+            {
+                byte[] bytes = _arrayPool.Rent(_blockSize);
+                try
+                {
+                    int offset = 0;
+                    do
+                    {
+                        if (async)
+                        {
+                            read = await stream.ReadAsync(bytes, offset, _blockSize - offset, cancellationToken).ConfigureAwait(false);
+                        }
+                        else
+                        {
+                            cancellationToken.ThrowIfCancellationRequested();
+                            read = stream.Read(bytes, offset, _blockSize - offset);
+                        }
+                        offset += read;
+                        absolutePosition += read;
+                    } while (offset < _minimumBlockFill && read != 0);
```

Comment:
What is `_minimumBlockFill` achieving?  Letting us do chunkier disk I/O and send network requests out faster?

Reply:
It aggregates the input stream read and makes sure we don't send blocks that are too small.